transcoder/src/storage/storage.go

```diff
@@ -0,0 +1,74 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+type ContentsWriterCallback func(ctx context.Context, writer io.Writer) error
+
+// Handles storage and retrieval of items in a storage backend.
+// "Items" are pieces of data that can be stored and retrieved.
+// Paths with ".." are not allowed, and may result in unexpected behavior.
+type StorageBackend interface {
+	// DoesItemExist checks if an item exists in the storage backend.
+	DoesItemExist(ctx context.Context, path string) (bool, error)
+	// ListItemsWithPrefix returns a list of items in the storage backend that match the given prefix.
+	// Note: returned items may have a "/" in them, e.g. "foo/bar/baz".
+	ListItemsWithPrefix(ctx context.Context, pathPrefix string) ([]string, error)
+	// DeleteItem deletes an item from the storage backend. If the item does not exist, it returns nil.
+	// Deletion should be "syncronous" (i.e. the function should block until the write is complete).
+	DeleteItem(ctx context.Context, path string) error
+	// SaveItemWithCallback saves an item to the storage backend. If the item already exists, it overwrites it.
+	// The writeContents function is called with a writer to write the contents of the item.
+	// Writes should be "syncronous" (i.e. the function should block until the write is complete).
+	SaveItemWithCallback(ctx context.Context, path string, writeContents ContentsWriterCallback) error
+	// GetItem retrieves an item from the storage backend.
+	GetItem(ctx context.Context, path string) (io.ReadCloser, error)
+}
+
+// SaveFilesToBackend is a helper function that saves files from the source directory to the backend storage.
+func SaveFilesToBackend(ctx context.Context, backend StorageBackend, sourceDirectory, destinationBasePath string) error {
+	if err := filepath.WalkDir(sourceDirectory, func(sourcePath string, d fs.DirEntry, err error) error {
```

Comment:
I'm putting the comment here but this is true for other lines/files as well:

I think `if err := function(); err != nil` is a good pattern ONLY when `function()` is short. When `function()` is multiline I find it harder to read & i'd rather we have

```go
err := function(
	with,
	multiples,
	args,
)
if err != nil {}
```